File: 1105-filling-bookcase-shelves/1105-filling-bookcase-shelves.py
class Solution(object):
    def minHeightShelvesHelper(self, idx, remainingWidth, currentShelfHeight, shelfWidth, books, memo):
        if(idx == len(books)):
            return currentShelfHeight

        if (idx, remainingWidth, currentShelfHeight) in memo:
            return memo[(idx, remainingWidth, currentShelfHeight)]

        bookWidth = books[idx][0]
        bookHeight = books[idx][1]

        placeInCurrentShelf = 1e9
        if(bookWidth <= remainingWidth) :
            placeInCurrentShelf = self.minHeightShelvesHelper(idx + 1, remainingWidth - bookWidth, max(currentShelfHeight, bookHeight), shelfWidth, books, memo)
        

        placeInNextShelf = currentShelfHeight + self.minHeightShelvesHelper(idx + 1, shelfWidth - bookWidth, bookHeight, shelfWidth, books, memo)

        memo[(idx, remainingWidth, currentShelfHeight)] = min(placeInCurrentShelf, placeInNextShelf)
        return memo[(idx, remainingWidth, currentShelfHeight)]
        

    def minHeightShelves(self, books, shelfWidth):
        """
        :type books: List[List[int]]
        :type shelfWidth: int
        :rtype: int
        """
        n = len(books)
        memo = {}
        return self.minHeightShelvesHelper(0, shelfWidth, 0, shelfWidth, books, memo)

        

File: 1105-filling-bookcase-shelves/1105-filling-bookcase-shelves.cpp
class Solution {
private:
    int minHeightShelvesHelper(int idx, int remainingWidth, int currentShelfHeight, int shelfWidth, vector<vector<int>>& books, vector<vector<int>>& dp) {
        if (idx == books.size()) return currentShelfHeight;

        if (dp[idx][remainingWidth] != -1) return dp[idx][remainingWidth];

        int bookWidth = books[idx][0];
        int bookHeight = books[idx][1];

        // Case 1: Placing the book in the current shelf
        int placeInCurrentShelf = INT_MAX;
        if (bookWidth <= remainingWidth) {
            placeInCurrentShelf = minHeightShelvesHelper(idx + 1, remainingWidth - bookWidth, max(currentShelfHeight, bookHeight), shelfWidth, books, dp);
        }

        // Case 2: Placing the book in a new shelf
        int placeInNewShelf = currentShelfHeight + minHeightShelvesHelper(idx + 1, shelfWidth - bookWidth, bookHeight, shelfWidth, books, dp);

        return dp[idx][remainingWidth] = min(placeInCurrentShelf, placeInNewShelf);
    }

public:
    int minHeightShelves(vector<vector<int>>& books, int shelfWidth) {
        int n = books.size();
        vector<vector<int>> dp(n, vector<int>(shelfWidth + 1, -1));
        return minHeightShelvesHelper(0, shelfWidth, 0, shelfWidth, books, dp);
    }
};
